bugsnag-android-core/src/test/java/com/bugsnag/android/JournaledDocumentStressTest.kt

```diff
@@ -0,0 +1,145 @@
+package com.bugsnag.android
+
+import com.bugsnag.android.internal.JournaledDocument
+import com.bugsnag.android.internal.JsonHelper
+import org.junit.Assert
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import java.io.File
+import java.lang.Thread
+
+class JournaledDocumentStressTest {
+    @Rule
+    @JvmField
+    var folder: TemporaryFolder = TemporaryFolder()
+
+    @Test
+    fun testOverflow() {
+        val baseDocumentPath = folder.newFile("mydocument")
+        val bufferSize = 1000L
+        val entryCount = 50000
+        val expectedDocument = mutableMapOf<String, Any>()
+        for (i in 1..entryCount) {
+            expectedDocument["a$i"] = "x"
+        }
+        val document = JournaledDocument(
+            baseDocumentPath,
+            JournaledDocumentTest.standardType,
+            JournaledDocumentTest.standardVersion,
+            bufferSize,
+            bufferSize,
+            mutableMapOf()
+        )
+
+        val modifyThread = Thread {
+            for (i in 1..entryCount) {
+                document.addCommand("a$i", "x")
+            }
+        }
+        modifyThread.start()
+
+        var value: Any? = null
+        var i = 0
+        val maxI = entryCount / 10
+
+        while (modifyThread.state != Thread.State.TERMINATED) {
+            document.forEach {
+                // Just cycle through all entries
+            }
+            value = document["a$i"]
+            i = (i + 1) % maxI
+        }
+
+        // "use" value so that it doesn't get optimized away
+        print(value)
+
+        // Leave the OS time to flush the mem mapped file
+        Thread.sleep(100)
```

Comment:
Ideally this should be smarter - either there should be some polling to avoid waiting for 100ms each time, or there should be some way of detecting when the journal document has been modified and the entries can be read.

Reply:
I'm not sure how to make this smarter... Since it's a memory mapped file, we have no control over when it gets flushed to disk. I could use the modification time as a proxy, but then we run into race conditions where the flush happens before I sample the modification time, causing a flaky test. I can't just sample at the start because there could be any number of flushes during the course of the test and we can't be sure that the last modification was or was not flushed by the time the test verifies the file contents - it's OS controlled, which is why we're using it.

At this point in the test, it's making sure that an unclosed file contains what we expect it to, such that a crash before closing gives deterministic results (which we need or else we'll get corrupted data on relaunch). 100ms gives adequate time for the OS to flush, while not adding significant delays to the test.

Reply:
Polling every ~5ms for a max timeout of ~2 seconds seems like it would do the trick? Waiting for 100ms wastes time unnecessarily but also might not be enough if I/O is slow - leading to flakes.

Reply:
With the following modifications:

```kotlin
        while (modifyThread.state != Thread.State.TERMINATED) {
            document.forEach { _ ->
                // Just cycle through all entries
            }
            value = document["a$i"]
            i = (i + 1) % maxI
        }

        val baseline = document.snapshotPath.lastModified()

        for (i in 1..400) {
            if (document.snapshotPath.lastModified() != baseline) {
                break
            }
            Thread.sleep(5)
        }
```

Every test run duration was 2.6 seconds because the sync-to-disk had already happened before taking the baseline.

Disk I/O is to solid state storage, which even in the worst case will write the 500 byte block of data in far less than a millisecond. The 100ms pause was just to make absolutely sure that the OS has had so much time to schedule the write that it's impossible it didn't happen (it typically writes immediately, but can sometimes take a few milliseconds to do so on slower machines).